### PR TITLE
sdk/go: Simplify await(), prevent data races

### PR DESCRIPTION
**Pull request by AaronFriel**
Moves accessing all of the fields to be guarded by the lock.

### CODE REVIEW

File: sdk/go/pulumi/types.go [hunk 1]
```diff
@@ -238,41 +238,49 @@ func (o *OutputState) reject(err error) {
...
 		// If the result is an Output, await it in turn.
 		//
 		// NOTE: this isn't exactly type safe! The element type of the inner output really needs to be assignable to
 		// the element type of the outer output. We should reconsider this.
-		if ov, ok := o.value.(Output); ok {
+		if ov, ok := value.(Output); ok {
 			o = ov.getState()
 		} else {
-			return o.value, true, secret, deps, nil
+			return value, known, secret, deps, nil
```
Comment (AaronFriel) [inline sdk/go/pulumi/types.go:283]: Changing this to return `known` is a no-op as known must be true, but it's cleaner to repeat the identifier.

File: sdk/go/pulumi/types.go [hunk 2]
```diff
@@ -238,41 +238,49 @@ func (o *OutputState) reject(err error) {
...
-		}
-
-		o.cond.L.Lock()
-		for o.state == outputPending {
-			if ctx.Err() != nil {
-				return nil, true, false, nil, ctx.Err()
-			}
-			o.cond.Wait()
-		}
-		o.cond.L.Unlock()
```
Comment (t0yv0) [inline sdk/go/pulumi/types.go:259]: So in fact this wouldn't unlock if ctx.Err() != nil branch returned first? Sounds like we always want to unlock. Good!

**Review by t0yv0 — APPROVED** [review]
LGTM it looks to me that it does not change behavior but makes it cleaner to read.

**Comment by AaronFriel** [comment]
bors merge

**Pull request merged**